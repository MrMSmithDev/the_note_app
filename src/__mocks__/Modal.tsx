import React from 'react';

interface MockModalProps {
  children: React.ReactNode;
}

export default function Modal({ children }: MockModalProps) {
  return <div data-testid="mock-modal">{children}</div>;
}
