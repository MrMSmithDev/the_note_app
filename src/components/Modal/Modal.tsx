import { useEffect, useRef } from 'react';

export interface ModalProps {
  children: React.ReactNode;
  closeModal: () => void;
  btnRef: React.RefObject<HTMLButtonElement>;
}

const Modal: React.FC<ModalProps> = ({ btnRef, closeModal, children }) => {
  const ref = useRef(null);

  useEffect(() => {
    setTimeout(() => {
      if (ref.current) {
        function closeModalCallBack(e: MouseEvent) {
          const modal = ref.current;
          const btn = btnRef.current;

          if (
            modal &&
            !modal.contains(e.target as Node) &&
            btn &&
            !btn.contains(e.target as Node)
          ) {
            closeModal();
            document.removeEventListener('click', closeModalCallBack, true);
          }
        }

        document.addEventListener('click', closeModalCallBack, true);
        return () => document.removeEventListener('click', closeModalCallBack, true);
      }
    });
  });

  return (
    <div className="p-1 fixed top-0 left-0 right-0 bottom-0 bg-[rgba(0,0,0,0.3)] flex justify-center items-center animate-fade-in">
      <div ref={ref} className="modal-children">
        {children}
      </div>
    </div>
  );
};

export default Modal;
