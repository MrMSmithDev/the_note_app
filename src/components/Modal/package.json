{
    "name": "modal",
    "version": "1.0.0",
    "main": "Modal.tsx",
    "dependencies": {
      "react": "^19.0.0"
    }
  }
  