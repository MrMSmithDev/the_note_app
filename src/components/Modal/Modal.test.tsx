import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Modal from './Modal';
import React from 'react';

describe('Modal component', () => {
  let closeModal: jest.Mock;
  let btnRef: React.RefObject<HTMLButtonElement>;
  const mockChild: React.ReactNode = <p>Test Modal Child</p>;

  beforeEach(() => {
    closeModal = jest.fn();
    btnRef = { current: document.createElement('button') };
  });

  it('renders children correctly', () => {
    render(
      <Modal closeModal={closeModal} btnRef={btnRef}>
        {mockChild}
      </Modal>
    );

    expect(screen.getByText('Test Modal Child')).toBeInTheDocument();
  });

  it('calls closeModal when clicking outside', async () => {
    render(
      <div data-testid="Outer">
        <button ref={btnRef}>Open Modal</button>
        <Modal closeModal={closeModal} btnRef={btnRef}>
          {mockChild}
        </Modal>
      </div>
    );

    await waitFor(() =>
      expect(screen.getByText('Test Modal Child')).toBeInTheDocument()
    );

    fireEvent.click(screen.getByTestId('Outer'));

    await waitFor(() => expect(closeModal).toHaveBeenCalled());
  });

  it('does not call closeModal when clicking inside', async () => {
    render(
      <Modal closeModal={closeModal} btnRef={btnRef}>
        {mockChild}
      </Modal>
    );

    await waitFor(() =>
      expect(screen.getByText('Test Modal Child')).toBeInTheDocument()
    );

    fireEvent.click(screen.getByText('Test Modal Child'));

    await waitFor(() => expect(closeModal).not.toHaveBeenCalled());
  });
});
