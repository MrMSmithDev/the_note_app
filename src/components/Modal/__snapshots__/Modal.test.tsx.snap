// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders correctly to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-1 fixed top-0 left-0 right-0 bottom-0 bg-[rgba(0,0,0,0.3)] flex justify-center items-center animate-fade-in"
  >
    <div
      class="modal-children"
    >
      <p>
        Test Modal Child
      </p>
    </div>
  </div>
</DocumentFragment>
`;
